### PR TITLE
[mccall_fitted_vfi] JAX conversion 

### PR DESCRIPTION
**Pull request by longye-tian**
This PR updates the Job Search III (Fitted Value Function Iteration) lecture to use JAX instead of NumPy/Numba, following modern QuantEcon standards and fixing style guide violations.

### Style Fixes (per QuantEcon style guide)
- Fixed section heading capitalization (Title Rule #2):
  - `## The Algorithm` → `## The algorithm`
  - `### Value Function Iteration` → `### Value function iteration`  
  - `### Fitted Value Function Iteration` → `### Fitted value function iteration`

### JAX Conversion
- Replaced NumPy/Numba imports with JAX equivalents
- Converted `@jitclass` to `NamedTuple` with factory function pattern
- Updated random number generation to use JAX's PRNG keys
- Replaced imperative loops with `jax.lax.while_loop`
- Implemented functional programming patterns in exercises using `jax.vmap`
- Updated all array operations from NumPy to JAX

### Text Consistency
- Updated documentation references from `numpy.interp` to `jnp.interp`
- Fixed code references: `np.inf` → `jnp.inf`, `np.linspace` → `jnp.linspace`

### CODE REVIEW

File: lectures/mccall_fitted_vfi.md [hunk 1]
```diff
@@ -317,21 +317,18 @@ Is the impact on the reservation wage as you expected?
 Here is one solution
 
 ```{code-cell} python3
-mcm = McCallModelContinuous()
-mu_vals = np.linspace(0.0, 2.0, 15)
-w_bar_vals = np.empty_like(mu_vals)
-
-fig, ax = plt.subplots()
+def compute_res_wage_given_mu(μ):
```
Comment (HumphreyYang) [inline lectures/mccall_fitted_vfi.md]:
```suggestion
def compute_res_wage_given_μ(μ):
```

File: lectures/mccall_fitted_vfi.md [hunk 2]
```diff
@@ -317,21 +317,18 @@ Is the impact on the reservation wage as you expected?
...
+    model = create_mccall_model(μ=μ)
+    w_bar = compute_reservation_wage(model)
+    return w_bar
 
-for i, m in enumerate(mu_vals):
-    mcm.w_draws = lognormal_draws(μ=m)
-    w_bar = compute_reservation_wage(mcm)
-    w_bar_vals[i] = w_bar
+mu_vals = jnp.linspace(0.0, 2.0, 15)
```
Comment (HumphreyYang) [inline lectures/mccall_fitted_vfi.md]:
```suggestion
μ_vals = jnp.linspace(0.0, 2.0, 15)
```

File: lectures/mccall_fitted_vfi.md [hunk 3]
```diff
@@ -305,7 +305,7 @@ The exercises ask you to explore the solution and how it changes with parameters
 Use the code above to explore what happens to the reservation wage when the wage parameter $\mu$
 changes.
 
-Use the default parameters and $\mu$ in `mu_vals = np.linspace(0.0, 2.0, 15)`.
+Use the default parameters and $\mu$ in `mu_vals = jnp.linspace(0.0, 2.0, 15)`.
```
Comment (HumphreyYang) [inline lectures/mccall_fitted_vfi.md]:
```suggestion
Use the default parameters and $\mu$ in `μ_vals = jnp.linspace(0.0, 2.0, 15)`.
```

File: lectures/mccall_fitted_vfi.md [hunk 4]
```diff
@@ -368,23 +365,26 @@ Now compute it.  Is this as you expected?
...
+def compute_res_wage_given_s(s, m=2.0, seed=1234):
     a, b = m - s, m + s
-    mcm.w_draws = np.random.uniform(low=a, high=b, size=10_000)
-    w_bar = compute_reservation_wage(mcm)
-    w_bar_vals[i] = w_bar
+    key = jax.random.PRNGKey(seed)
+    uniform_draws = jax.random.uniform(key, shape=(10_000,), minval=a, maxval=b)
+    # Create model with default parameters but replace wage draws
+    model = create_mccall_model()
+    model = model._replace(w_draws=uniform_draws)
```
Comment (HumphreyYang) [inline lectures/mccall_fitted_vfi.md]: This is an interesting and smart workaround :thinking: 

But we are effectively creating one model and immediately replace it since `NamedTuple` is immutable.

I think a better pattern would be to define `NamedTuple` factory like this:

```
def create_mccall_model(c=1, α=0.1, β=0.96, grid_min=1e-10, grid_max=5, 
                        grid_size=100, μ=2.5, σ=0.5, mc_size=1000, 
                        seed=1234, w_draws=None):
    """Factory function to create a McCall model instance."""
    if w_draws is None:
        w_draws = lognormal_draws(n=mc_size, μ=μ, σ=σ, seed=seed)
    
    w_grid = jnp.linspace(grid_min, grid_max, grid_size)
    return McCallModelContinuous(c=c, α=α, β=β, w_grid=w_grid, w_draws=w_draws)
```

and then call

```
model = create_mccall_model(w_draws=uniform_draws)
```

here.

File: lectures/mccall_fitted_vfi.md [hunk 5]
```diff
@@ -175,123 +176,122 @@ plt.show()
...
+    def body_fun(state):
+        v, d, i, error = state
+        v_new, d_new = update(model, v, d)
+        error_1 = jnp.max(jnp.abs(v_new - v))
+        error_2 = jnp.abs(d_new - d)
+        error = jnp.maximum(error_1, error_2)
+        return v_new, d_new, i + 1, error
+    
+    def cond_fun(state):
+        v, d, i, error = state
```
Comment (HumphreyYang) [inline lectures/mccall_fitted_vfi.md]:
```suggestion
        _, _, i, error = state
```

File: lectures/mccall_fitted_vfi.md (hunk 4 above)
Comment (longye-tian) [reply lectures/mccall_fitted_vfi.md]: Thank you very much @HumphreyYang for your detailed review. I've updated the code accordingly.

**Comment by mmcky** [comment]
Thanks @longye-tian for an excellent PR, thanks @HumphreyYang for comments. I just had a few questions. 

1. the figure in `exercise 2` is slightly different. This might be OK -- but just thought I would check if this is due to a random seed, or if it has any implications for the exercise (LHS = this PR, RHS = Live site)

Screenshot 2025-08-21 at 4 50 25 pm

~~2. The timing is slower on JAX version (GitHub actions build - `gpu`) than the numba version. (LHS = this PR, RHS = Live site)~~ (**Status:** Local testing shows `jax` should run approximately 2 x faster than `numba` version, so this timing is an issue with GA environment). https://github.com/QuantEcon/lecture-python.myst/pull/567#issuecomment-3212869759

Screenshot 2025-08-21 at 4 53 16 pm

- [x] review `collab` comments to either agree or resolve.

**Comment by mmcky** [comment]
@HumphreyYang I see [95f5ef6](https://github.com/QuantEcon/lecture-python.myst/pull/567/commits/95f5ef672fdb826c44ec688ddc8963b14c2ef6d1) the result is 10s so 2s faster than `gpu` and is the same as the `numba` implementation.

**Comment by HumphreyYang** [comment]
> @HumphreyYang I see [95f5ef6](https://github.com/QuantEcon/lecture-python.myst/pull/567/commits/95f5ef672fdb826c44ec688ddc8963b14c2ef6d1) the result is 10s so 2s faster than `gpu` and is the same as the `numba` implementation.

Hi @mmcky, 

This is an interesting behaviour. I am curious could it be because of the `jb build` issue. 

I will run some more benchmark.

**Comment by mmcky** [comment]
> > @HumphreyYang I see [95f5ef6](https://github.com/QuantEcon/lecture-python.myst/pull/567/commits/95f5ef672fdb826c44ec688ddc8963b14c2ef6d1) the result is 10s so 2s faster than `gpu` and is the same as the `numba` implementation.
> 
> Hi @mmcky,
> 
> This is an interesting behaviour. I am curious could it be because of the `jb build` issue.
> 
> I will run some more benchmark.

So I ran the `main` branch and this branch locally

| state | time |
|------|------|
| main (numba) | 6s |
| job_separation (jax) | 3s |

so it does appear to be a GitHub Actions thing. 
Surprising as we are running this on an `g4dn.2xlarge` which has 8 vCPUs with 64Gb RAM.

**Comment by HumphreyYang** [comment]
Hi @mmcky,

Please find the gist result from my end:

https://gist.github.com/HumphreyYang/a488df0e8ef01b8d91af6e6309d5e45f

It should run around 2x faster than the numba version in both CPU and GPU.

There are while loops in the code so I expect GPU wouldn't help much.

**Comment by HumphreyYang** [comment]
> > > @HumphreyYang I see [95f5ef6](https://github.com/QuantEcon/lecture-python.myst/pull/567/commits/95f5ef672fdb826c44ec688ddc8963b14c2ef6d1) the result is 10s so 2s faster than `gpu` and is the same as the `numba` implementation.
> > 
> > 
> > Hi @mmcky,
> > This is an interesting behaviour. I am curious could it be because of the `jb build` issue.
> > I will run some more benchmark.
> 
> So I ran the `main` branch and this branch locally
> 
> state	time
> main (numba)	6s
> job_separation (jax)	3s
> so it does appear to be a GitHub Actions thing. Surprising as we are running this on an `g4dn.2xlarge` which has 8 vCPUs with 64Gb RAM.

Great! Our number matches! It looks like 8vCPUs are not helping : )

**Comment by mmcky** [comment]
@longye-tian I have updated the timing comment above (with strikethrough) based on our investigations and opened https://github.com/QuantEcon/lecture-python.myst/issues/568

**Comment by HumphreyYang** [comment]
> 1. the figure in `exercise 2` is slightly different. This might be OK -- but just thought I would check if this is due to a random seed, or if it has any implications for the exercise (LHS = this PR, RHS = Live site)

Thanks @mmcky for the detailed review:

JAX and NumPy use different pseudorandom number generators (PRNGs), so even with the same seed, they produce different random numbers. Below is an illustration:
https://gist.github.com/HumphreyYang/f8aa7b322c48723bac568290de8a08fc

**Comment by mmcky** [comment]
> > 1. the figure in `exercise 2` is slightly different. This might be OK -- but just thought I would check if this is due to a random seed, or if it has any implications for the exercise (LHS = this PR, RHS = Live site)
> 
> Thanks @mmcky for the detailed review:
> 
> JAX and NumPy use different pseudorandom number generators (PRNGs), so even with the same seed, they produce different random numbers. Below is an illustration: https://gist.github.com/HumphreyYang/f8aa7b322c48723bac568290de8a08fc

Thanks @HumphreyYang so once we change to `jax` that graph will become static again (with the new shape)? Also just to confirm, the new shape doesn't change the interpretation in any way re: context of the solution?

**Comment by HumphreyYang** [comment]
> > > 1. the figure in `exercise 2` is slightly different. This might be OK -- but just thought I would check if this is due to a random seed, or if it has any implications for the exercise (LHS = this PR, RHS = Live site)
> > 
> > 
> > Thanks @mmcky for the detailed review:
> > JAX and NumPy use different pseudorandom number generators (PRNGs), so even with the same seed, they produce different random numbers. Below is an illustration: https://gist.github.com/HumphreyYang/f8aa7b322c48723bac568290de8a08fc
> 
> Thanks @HumphreyYang so once we change to `jax` that graph will become static again (with the new shape)? Also just to confirm, the new shape doesn't change the interpretation in any way re: context of the solution?

Hi @mmcky, yes, the new PRNGs should output the same solution everytime and the interpretation is not changed. Everything is in the same magnitude as the old code so the interpretations are the same.

File: lectures/mccall_fitted_vfi.md [hunk 6]
```diff
@@ -368,23 +372,25 @@ Now compute it.  Is this as you expected?
...
-
-for i, s in enumerate(s_vals):
+def compute_res_wage_given_s(s, m=2.0, seed=1234):
     a, b = m - s, m + s
-    mcm.w_draws = np.random.uniform(low=a, high=b, size=10_000)
-    w_bar = compute_reservation_wage(mcm)
-    w_bar_vals[i] = w_bar
+    key = jax.random.PRNGKey(seed)
+    uniform_draws = jax.random.uniform(key, shape=(10_000,), minval=a, maxval=b)
+    # Create model with default parameters but replace wage draws
```
Comment (mmcky) [inline lectures/mccall_fitted_vfi.md:382]:
```suggestion

    # Create model with default parameters but replace wage draws
```

File: lectures/mccall_fitted_vfi.md [hunk 7]
```diff
@@ -368,23 +372,25 @@ Now compute it.  Is this as you expected?
...
+    model = create_mccall_model(w_draws=uniform_draws)
+    w_bar = compute_reservation_wage(model)
+    return w_bar
 
+s_vals = jnp.linspace(1.0, 2.0, 15)
+# Use vmap with different seeds for each s value
```
Comment (mmcky) [inline lectures/mccall_fitted_vfi.md:389]:
```suggestion

# Use vmap with different seeds for each s value
```

File: lectures/mccall_fitted_vfi.md [hunk 8]
```diff
@@ -175,123 +179,126 @@ plt.show()
...
+                        grid_max=5,
+                        grid_size=100,
+                        μ=2.5,
+                        σ=0.5,
+                        mc_size=1000,
+                        seed=1234,
+                        w_draws=None):
+    """Factory function to create a McCall model instance."""
+    if w_draws is None:
+        # Generate wage draws if not provided
```
Comment (mmcky) [inline lectures/mccall_fitted_vfi.md:221]:
```suggestion

        # Generate wage draws if not provided
```

File: lectures/mccall_fitted_vfi.md [hunk 9]
```diff
@@ -175,123 +179,126 @@ plt.show()
...
+        w_draws = lognormal_draws(n=mc_size, μ=μ, σ=σ, seed=seed)
+
+    w_grid = jnp.linspace(grid_min, grid_max, grid_size)
+    return McCallModelContinuous(c=c, α=α, β=β, w_grid=w_grid, w_draws=w_draws)
+
+@jax.jit
+def update(model, v, d):
+    """Update value function and continuation value."""
+    # Unpack model parameters
```
Comment (mmcky) [inline lectures/mccall_fitted_vfi.md:231]:
```suggestion

    # Unpack model parameters
```

File: lectures/mccall_fitted_vfi.md [hunk 10]
```diff
@@ -175,123 +176,126 @@ plt.show()
...
-
+    c, α, β, w_grid, w_draws = model
+    u = jnp.log
+    
+    v, d = solve_model(model)
+    h = u(c) + β * d
+    
+    # Find the first wage where v(w) >= h
+    indices = jnp.where(v >= h, size=1, fill_value=-1)
+    w_bar = jnp.where(indices[0] >= 0, w_grid[indices[0]], jnp.inf)
```
[the remark this answers is not in the document]
Comment (mmcky) [reply lectures/mccall_fitted_vfi.md:302]: not merging this suggestion.

**Comment by mmcky** [comment]
thanks @longye-tian for an excellent PR. 

thanks @HumphreyYang for your comments and reviews. 

@jstac this will serve as another reference in the `numba` -> `jax` conversion work.

**Comment by mmcky** [comment]
@jstac I tagged your review for this one so you have 👀 on what the team has put together.

I'll merge this on Monday.
